### PR TITLE
[SNAP-2054][SNAP-2055] simplify delta encoding, fix broadcast caching

### PR DESCRIPTION
**Pull request by sumwale**
This PR addresses the issues SNAP-2054 and SNAP-2055 seen when simplifying the delta
encoder APIs to avoid sorting encoded deltas on positions which required re-arranging
and used to work on top of "unfinished" encoder. The decoders on top of "unfinished" encoders
were added for this purpose to allow decoding without encoder having created the final buffer.
This was unnecessary because for most cases where there is no intermediate
exchange/broadcast then output from ColumnTableScan is already in order of positions in the
batch, while for the exchange/broadcast case the sort on batchId could as well also do sort
on ordinal and avoid this complication.

The change and tests revealed two issues (three if one also counts the BroadcastNestedLoopJoin
  caching issue) which are also fixed in this PR.

## Changes proposed in this pull request

- using the API as done in SNAP-2033 to ensure sorting on (batchId, ordinal) instead of batchId
- this ensures that positions within a batch are already sorted without requiring the delta
  encoder to re-arrange the deltas in sorted order; the ColumnTableScan itself is marked as
  providing the output rows already sorted by (batchId, ordinal)
- removed "decoderBeforeFinish" and related APIs simplifying the code with above change
- boolean bitset encoder works a bit differently where the returned "cursor" is not actually
  position in data bytes so using appropriate calls to get the data offset; overriding "offset"
  method in boolean encoder for the same
- add NO_CACHING for BroadcastNestedLoopJoin and also if there are no ParamLiterals
- added tests for above and updated cache size expectation in TokenizationTest

## Patch testing

precheckin

## ReleaseNotes.txt changes

NA

## Other PRs 

NA

### CODE REVIEW

**Comment by rishitesh** [comment]
@sumwale What is an unfinished encoder ?

**Comment by sumwale** [comment]
@rishitesh It means without using the final buffer produced by ColumnEncoder.finish rather reading the raw encoded data in the encoder. For example for dictionary encoder, the dictionary will be a separate buffer and indexes in separate buffer which will be joined together in finish. Likewise nullability bitmask is tracked in separate array (see NullableEncoder) and joined to copied to final buffer in finish.

The normal decoders work only on top of "finished" buffers so the code earlier in delta encoder had changed decoders to be able to also handle "unfinished" encoders (e.g. allow for separate dictionary and index buffers). This has been removed now since there is no longer a need to read directly from encoders.

**Comment by rishitesh** [comment]
@sumwale "sorting encoded deltas on positions" . This was required only for merging of delta  or some other requirement ?

**Comment by sumwale** [comment]
@rishitesh also when reading. The decoder will look for next updated/deleted position, and when that position is reached then read from update/delete delta instead of main buffer. So it has to be sorted so that decoder can read next position else for every read decoder will have to search through the entire delta to check if its updated/deleted.

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala [hunk 1]
```diff
@@ -561,34 +546,27 @@ final class ColumnDeltaEncoder(val hierarchyDepth: Int) extends ColumnEncoder {
...
     realEncoder.setSource(buffer, releaseOld = false)
     val columnBytes = allocator.baseObject(buffer)
     var cursor = allocator.baseOffset(buffer)
     // write the header including positions
     cursor = writeHeader(columnBytes, cursor, numNullWords, numBaseRows,
-      positionsArray, startIndex, endIndex)
+      positionsArray, numDeltas)
 
     // write any internal structures (e.g. dictionary)
     cursor = realEncoder.writeInternals(columnBytes, cursor)
```
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala]: Can't we call realEncoder.finish() here, which will write any internal structures ?

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala [hunk 2]
```diff
@@ -532,27 +530,14 @@ final class ColumnDeltaEncoder(val hierarchyDepth: Int) extends ColumnEncoder {
...
-    val startIndex = if (numDeltas <= 1) 0
-    else RadixSort.sort(positions, numDeltas, 0, 3, false, false)
-    val endIndex = startIndex + numDeltas
-    val decoder = realEncoder.decoderBeforeFinish(encoderCursor)
-    val decoderBuffer = realEncoder.buffer
-    val decoderNullBytes = realEncoder.getNullWords
-    val decoderData = realEncoder.columnData
-    val decoderAllocator = realEncoder.allocator
+    // write any remaining bytes in encoder
+    val dataEndPosition = realEncoder.flushWithoutFinish(encoderCursor)
```
[the remark this answers is not in the document]
Comment (sumwale) [reply core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala]: We no longer require sorting on index positions, but still need to copy the updated positions + values into the final exact size buffer. Notice the difference in layouts:

normal encoder finish: header + null bitmask + data

delta encoder: header + null bitmask + updated positions + data

Moving this inside finish will mean two copies, one in the call to finish and then copy from finished buffer to update delta (+ will require methods to know the boundaries in finished buffer).

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala (hunk 1 above)
Comment (sumwale) [reply core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala]: Avoids one additional copy in finish. Even with finish another methods will be needed to determine the boundary where data starts to create the new delta layout as mentioned above (see the writeHeader method that also writes the updated positions).

File: core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala [hunk 3]
```diff
@@ -561,34 +546,27 @@ final class ColumnDeltaEncoder(val hierarchyDepth: Int) extends ColumnEncoder {
...
+    val cursorOffset = realEncoder.offset(cursor)
+    cursor = realEncoder.columnBeginPosition + cursorOffset
+
+    // copy the encoded data
+    assert(cursor + dataSize == realEncoder.columnEndPosition)
+    Platform.copyMemory(dataBuffer, dataBeginPosition, columnBytes, cursor, dataSize)
 
     // release the old buffer
-    decoderAllocator.release(decoderData)
+    dataAllocator.release(dataColumnBytes)
```
[the remark this answers is not in the document]
Comment (sumwale) [reply core/src/main/scala/org/apache/spark/sql/execution/columnar/encoding/ColumnDeltaEncoder.scala]: Initialize has to be called in any case even for first use. It provides the dataType and tentative size.

**Pull request merged**